feat: add C impolementation for `stdlib/stats/base/dists/cauchy/quantile`

NishchayRajput opened the pull request:
Resolves #3486.

## Description

> What is the purpose of this pull request?

This pull request:

-   [RFC]: Add C implementation for `@stdlib/stats/base/dists/cauchy/quantile`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3486 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

NishchayRajput committed:
feat: add C impolementation for @stdlib/stats/base/dists/cauchy/quantile

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/README.md:
```diff
@@ -149,6 +149,115 @@ for ( i = 0; i < 10; i++ ) {
 
 <!-- /.examples -->
 
+<!-- /.examples -->
+
```
```suggestion
```

NishchayRajput replied:
done!

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/README.md:
```diff
@@ -149,6 +149,115 @@ for ( i = 0; i < 10; i++ ) {
+
+#### stdlib_base_dists_cauchy_quantile( p, x0, gamma )
+
+Evaluates the [quantile function][quantile-function] of a [Cauchy][cauchy-distribution] distribution with location parameter x0 and scale parameter gamma.
```
```suggestion
Evaluates the [quantile function][quantile-function] of a [Cauchy][cauchy-distribution] distribution with location parameter `x0` and scale parameter `gamma`.
```

NishchayRajput replied:
done!

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/README.md:
```diff
@@ -149,6 +149,115 @@ for ( i = 0; i < 10; i++ ) {
+
+The function accepts the following arguments:
+
+-   **p**: `[in] double` probability
```
```suggestion
-   **p**: `[in] double` probability.
```

NishchayRajput replied:
done!

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/README.md:
```diff
@@ -149,6 +149,115 @@ for ( i = 0; i < 10; i++ ) {
+}
+
+int main( void ) {
+    double p;
+    double gamma;
+    double x0;
+    double y;
```
```suggestion
    double gamma;
    double x0;
    double y;
    double p;
```

NishchayRajput replied:
I just saw on other PR a comment regarding the same that it should be order according to the length! and  for the same length is it then need to be ordered by alphabetically? is that so?

kgryte replied:
Length is good enough.

NishchayRajput replied:
done

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/README.md:
```diff
@@ -149,6 +149,115 @@ for ( i = 0; i < 10; i++ ) {
+        x0 = random_uniform( 0.0, 10.0 ) - 5.0;
+        gamma = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
+        y = stdlib_base_dists_cauchy_quantile( p, x0, gamma );
+        printf( "x0: %lf, gamma: %lf, F(x0;gamma): %lf\n", x0, gamma, y );
```
See L144

NishchayRajput replied:
done

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/examples/c/example.c:
```diff
@@ -0,0 +1,43 @@
+}
+
+int main( void ) {
+	double p;
+	double gamma;
+	double x0;
+	double y;
```
```suggestion
	double gamma;
	double x0;
	double y;
	double p;
```

NishchayRajput replied:
done

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/examples/c/example.c:
```diff
@@ -0,0 +1,43 @@
+		x0 = random_uniform( 0.0, 10.0 ) - 5.0;
+		gamma = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
+		y = stdlib_base_dists_cauchy_quantile( p, x0, gamma );
+		printf( "x0: %lf, gamma: %lf, F(x0;gamma): %lf\n", x0, gamma, y );
```
Same comment as README

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/include/stdlib/stats/base/dists/cauchy/quantile.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Returns the differential quantile of a Cauchy distribution.
```
This description is incorrect.

NishchayRajput replied:
done

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/src/main.c:
```diff
@@ -0,0 +1,48 @@
+#include "stdlib/constants/float64/pi.h"
+
+/**
+* Returns the differential entropy of a Cauchy distribution.
```
Description is incorrect.

NishchayRajput replied:
done

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/src/main.c:
```diff
@@ -0,0 +1,48 @@
+/**
+* Returns the differential entropy of a Cauchy distribution.
+*
+* @param p        input value
```
```suggestion
* @param p        input probability
```

NishchayRajput replied:
done

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/test/test.native.js:
```diff
@@ -0,0 +1,183 @@
+	t.end();
+});
+
+tape( 'the function evaluates the quantile function at `p` given `x0` and `gamma` ( large `gamma`)', opts, function test( t ) {
```
```suggestion
tape( 'the function evaluates the quantile function at `p` given `x0` and `gamma` (large `gamma`)', opts, function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/test/test.native.js:
```diff
@@ -0,0 +1,183 @@
+	var p;
+	var y;
+	var i;
+
```
Is there a reason why you removed the comment present in `test.quantile.js`?

NishchayRajput replied:
Sorry I wasn't aware of it! I just corrected now!

NishchayRajput committed:
fix: updated the reviewed changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

NishchayRajput commented:
@kgryte @Planeshifter I have made the suggested changes please look into it!

Planeshifter committed:
fix: correct typos and function name in C implementation for `stats/base/dists/cauchy/quantile`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
fix: use 'Evaluates' instead of 'Returns' in header comment

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/cauchy/quantile/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/stats/base/dists/cauchy/quantile.h"
+#include "stdlib/math/base/napi/ternary.h"
+
+// cppcheck-suppress shadowFunction
```
```suggestion
```

Planeshifter committed:
fix: update examples to use proper uniform distribution ranges

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
fix: update benchmarks to use proper uniform distribution ranges

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
fix: update README examples to use proper uniform distribution ranges

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: remove cppcheck directive

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
Merge branch 'develop' into nishchay-base-cauchy-quantile

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
Thanks, @NishchayRajput!
